🚀 소셜 로그인 (kakao login + jwt token) 구현

shoeone96 opened the pull request:
## 🎫 관련 이슈
<!--이슈 태스크를 모두 완료하고 닫는다면 * Resolves #번호-->
<!--이슈 태스크를 모두 완료하지는 못 했지만 닫는다면 * Closes #번호-->
<!--열어둔다면 * #번호-->

* Resolves #28 

## 🚀 주요 변경사항
- 카카오 기반 소셜 로그인 구현
- 다른 소셜 로그인이 추가되더라도 구현 가능하도록 OAuth2 기반으로 로직 작성
- 소셜 로그인 인증 후 인증 정보를 기반으로 user 정보를 생성
    - user 정보 저장(고유 id, provider, providerId 기반으로 식별할 수 있는 user 생성)
    - 이후 해당 user 정보를 기반으로 jwt token을 생성한 후 프론트에 넘겨주는 방식으로 구현
- filter에서 error 발생 시 `CustomAuthenticationEntryPoint`를 통해서 에러 식별 처리
    - objectMapper를 통해 ErrorResponse를 body에 담아 내려주는 방식으로 filter error 처리
        -> 프론트 엔드와 공통 에러 코드 규약 만든 것을 반영하기 위해 에러코드를 담을 수 있는 객체 내려주기
- 소셜 로그인 로직에 맞춘 validation/test 코드 작성

<!--빠른 리뷰를 위해 이해를 도울 만한 설명을..-->

## 💡 기타사항
추후 구현 예정
- 추가 정보 받아오는 api 구현
- refresh token 구현
- 로그아웃, 회원 탈퇴 구현
<!-- ex) 질문. 이후에 이런걸 할거고 또한 지금은 이러한 이유 때문에 이런걸 작업했다. 의존성, 추후해야할 일 등등-->

shoeone96 committed:
chore: 조건별 yml 파일 설정

shoeone96 committed:
chore: check style

shoeone96 committed:
chore: spring security, OAuth2, data faker 의존성 추가

shoeone96 committed:
feat: jwt 기본 설정

shoeone96 committed:
feat: security 기본 접근 설정

oauth2 설정, 에러 처리 설정, jwt 설정 추가

shoeone96 committed:
feat: property 설정값 인식 어노테이션 추가

shoeone96 committed:
feat: login user 정보 parameter로 받아오기 위한 처리

shoeone96 committed:
feat: security filter 내부에서 에러 처리하기 위한 Entry point 추가

shoeone96 committed:
feat: user login , join 서비스 로직 작성

shoeone96 committed:
feat: user 조회 및 유저 권한 목록 조회 기능 추가

shoeone96 committed:
feat: user 검증 로직 추가

shoeone96 committed:
feat: user 에러 코드 추가

shoeone96 committed:
feat: 소셜로그인 관련 validator 추가

shoeone96 committed:
feat: 잘못된 인증 관련 custom 예외 추가

shoeone96 committed:
feat: 로그인 완료 시 반환되는 responseDto 추가

shoeone96 committed:
feat: 소셜 로그인 성공 시 받아오는 값을 토큰으로 처리하는 로직 추가

shoeone96 committed:
feat: 소셜 로그인 기능 밴더사 구분하는 로직 추가

shoeone96 committed:
refactor: 공통 validator 수정

shoeone96 committed:
test: user service 로직 test

shoeone96 committed:
test: provider 반환 테스트 로직 추가

shoeone96 committed:
feat: jwt 토큰을 정상적으로 반환하는 로직 추가

shoeone96 committed:
fix: merge fixed

shoeone96 committed:
style: code reformat

shoeone96 committed:
style: key 값 대문자 통일

shoeone96 committed:
fix: build 터지는 문제, checkstyle, yml 파일 비밀 키 수정

IjjS commented on src/main/java/team/silvertown/masil/config/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,53 @@
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends GenericFilterBean {
```
GenericFilterBean을 상속하는 필터 중 OncePerRequestFilter라는 클래스가 있습니다. 요청마다 한 번만 거치는 걸 보장해주는 필터인데 그걸 사용하는건 어떨까요 형님?

ASak1104 replied:
다른 서블릿으로 dispatch 되면 중복 적용될 수 있으니 한 번만 적용한다.. 미쳤는데요 인재형님?

shoeone96 replied:
네 여러 reference를 통해 확인했습니다! 다른 api로 redirect 하는 경우 등에 다시 해당 filter를 거치는 중복이 발생하고 있네요!

이 부분 trouble shooting에도 반영해볼게요!

reference:
[Springboot OncePerRequestFilter 와 GenericFilterBean의 차이](https://dev-racoon.tistory.com/34)

IjjS commented on build.gradle:
```diff
@@ -42,8 +42,17 @@ dependencies {
     implementation 'org.hibernate.orm:hibernate-spatial'
 
+    // oauth2
+    implementation 'org.springframework.boot:spring-boot-starter-oauth2-client:3.2.0'
```
저희 부트는 3.2.2쓰는데 버전 맞추는 건 어떻게 생각하시나요??

IjjS commented on src/main/java/team/silvertown/masil/MasilApplication.java:
```diff
@@ -2,8 +2,10 @@
+import org.springframework.boot.context.properties.ConfigurationPropertiesScan;
 
 @SpringBootApplication
+@ConfigurationPropertiesScan
```
👍

IjjS commented on src/main/java/team/silvertown/masil/config/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,53 @@
+            Authentication authentication = tokenProvider.getAuthentication(token);
+
+            context.setAuthentication(authentication);
+        }
```
현재 토큰이 잘못됐을 때 필터 흐름은
JwtAuthentication -> RequestCache -> ... -> ExceptionTranslation -> AccessDenied -> ExceptionTranslation -> EntryPoint
이런식으로 가게 됩니다.

이 필터에서 예외 발생 시 바로 EntryPoint로 가게 하는 건 어떨까요?
예를 들면,
```java
public class JwtAuthenticationFilter {
    ...
    private CustomAuthenticationEntryPoint authenticationEntryPoint;
    ...
    public void doFilter(...) {
        try {
            토큰
        } catch (무슨 무슨 예외 e) {
            authenticationEntryPoint.commence(...);
        }
        ...
    }
}
```
이런 식으로나 try catch 대신 if문으로 무언갈 줘도 괜찮을 것 같고요.

어디까지나 이러면 좋지 않을까? 하는 방향이라 동작한다면 상관 없을 것 같긴 합니다. 의견주세요!

ASak1104 replied:
> try { 토큰 } catch (무슨 무슨 예외 e) { authenticationEntryPoint.commence(...); }

@Team-SilverTown/backend  스프링 기본 인증 필터에서도 저런식으로 구현을 많이 하나요?

토큰이 잘못되면 필터 체인에서 나중 애들을 모두 무시하고 바로 터뜨릴 것이냐 vs 토큰이 잘못돼도 로그인과 같은 api 콜을 위해 살려둘 것이냐 라고 생각이 드는데

형님들은 어떤 방식이 더 마음에 드세요?

IjjS replied:
토큰이 잘못된 것 != 토큰이 없는 것
토큰이 없을 때는 통과를 하고 provider를 거쳐 유효하지 않은 토큰만 쳐내는겁니다! 어떤가요?

ASak1104 replied:
유효하지 않은 토큰에 대해서도 로그인 같은 곳은 통과가 되야할 것 같은데 어떻게 생각하시나요 형님

IjjS replied:
상관없습니다 이대로 하시죠 그럼

shoeone96 replied:
로그인이 필요한 곳에서 jwt token을 넣을 때 

1. 토큰이 없을 때
2. 유효하지 않은 토큰일 때 

-> 바로 예외를 터뜨리는 건가 vs 기존 흐름으로 필터를 타게 만드는 게 나은가

어차피 유효하지 않은 토큰, 토큰이 null 인 경우 토큰 set을 안하고 setAuthentication이 안되면 
AnonymusAuthenticationToken 을 기본적으로 Authentication에 저장하는 형태가 될텐데 
그걸로 추후에 인증, 인가 검증을 하게 될테고 이 흐름을 놔두는 방향도 괜찮을 것 같다고 생각하긴 합니다!

IjjS commented on src/main/java/team/silvertown/masil/config/jwt/JwtProperties.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("jwt")
```
이러면 바로 constructor binding 적용 되는건가요? 대박

IjjS commented on src/main/java/team/silvertown/masil/config/jwt/TokenMessage.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class TokenMessage {
```
JwtTokenProvider에 private으로 안 넣고 따로 관리하신 이유가 있을까요??

ASak1104 replied:
@IjjS 빽엔두리에서 개발할 때 제가 이런 방식으로 구현해서 그렇습니다 형님... 메시지에 대해선 상수 클래스나 enum으로 관리했었거든요

shoeone96 replied:
그때 만들었던 게 다시 봐도 좋아서... jwt 부분은 현우 걸 거의 차용했어요

IjjS commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
👍

IjjS commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,67 @@
+        Authentication oauth2Authentication
+    ) throws ServletException, IOException {
+        Object principal = oauth2Authentication.getPrincipal();
+        log.info("OAuth2 authentication success: {}", principal);
```
로그를 info로 쌓을 필요는 없을 것 같습니다! trace나 debug로 하면 어떨까요? 프로덕션에선 info 로그가 나올텐데 누군가 로그인하면 로그가 계속 생성되게 되니까..

shoeone96 replied:
실제 구현하면서 체크해볼 때 계속 확인했던 건데 지금 지워도 될 것 같아요 지우는 건 어떤가요?

IjjS replied:
좋습니다

IjjS commented on src/main/java/team/silvertown/masil/config/security/CustomAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,35 @@
+@Configuration
+public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
빈 주입으로도 될 것 같은데 생성하신 이유가 있나요??

shoeone96 replied:
오.. objectMapper 사용할 일이 거의 없어서 자동 bean 등록이 되는줄 몰랐네요

[이런 식](https://sabarada.tistory.com/236)으로 등록되어 있네요

IjjS commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
💯 주석은 빼도 좋을 것 같습니다

IjjS commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+        http
+            .csrf(AbstractHttpConfigurer::disable)
+            .formLogin(AbstractHttpConfigurer::disable)
+            .headers(AbstractHttpConfigurer::disable)
```
HeaderWriterFilter였나요 이건? 있는 것과 없는 것의 차이가 뭔가요??

ASak1104 replied:
[이런 차이](https://kangwoojin.github.io/programing/spring-security-basic-header-wrtier-filter/)가 있다고 하네요
api 서버라 없어도 될 것 같은데 있는게 나으려나요?

IjjS replied:
없는거로 가시죠!

IjjS commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+                handler -> handler.authenticationEntryPoint(customAuthenticationEntryPoint));
+
+        //기타 설정 +  OAuth successHandler 설정
+        http
```
RequestCacheFilter는 disable을 안 해주셨는데 하면 안된다고 생각하신 이유가 있으신가요?

ASak1104 replied:
그렇네 해야겠네요

IjjS commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+            authorities.forEach((authority) -> {
+                GrantedAuthority mappedAuthority;
+
+                if (authority instanceof OidcUserAuthority userAuthority) {
```
저희는 OIDC를 아예 거칠 일이 없을 것 같은데 불필요한가 싶기도 합니다! 람다 함수도 분리하면 좋을 것 같지만 SecurityConfig니까 넘어가죠 ㅎㅎ

ASak1104 replied:
그렇긴 하네요

shoeone96 replied:
spring security 공식 설정에서 제안한 내용이라 커스텀해도 무관하긴 합니다!

IjjS commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+    private final JwtTokenProvider jwtTokenProvider;
+    private final JwtAuthenticationFilter jwtAuthenticationFilter;
+    private final ObjectMapper objectMapper;
+    private final CustomAuthenticationEntryPoint customAuthenticationEntryPoint;
```
OAuth2SuccessHandler도 생성자 주입으로 받으면 문제가 있으려나요??

shoeone96 replied:
어노테이션으로 클래스명 자체에 빈으로 생성된 게 아니라 securityConfig 내부에서 bean 설정을 하고 있긴 한데 이 부분은 @Configuration 등의 어노테이션을 붙여서 생성자 주입으로 넣어도 상관 없을 것 같긴 하네요

더 깔끔해보이는 방식 말씀해주시면 그에 맞춰 해볼게요

IjjS replied:
둘 다 괜찮아보이는데 편하신대로 하시죠! 지금도 좋습니다

IjjS commented on src/main/java/team/silvertown/masil/user/domain/Provider.java:
```diff
@@ -1,5 +1,27 @@
+    }
+
+    public static Provider returnProvider(String value) {
+        Provider[] values = Provider.values();
+        for (Provider provider : values) {
+            if (provider.getValue()
+                .equals(value)) {
+                return provider;
+            }
+        }
+        throw new InvalidAuthenticationException(UserErrorCode.INVALID_PROVIDER);
```
Arrays.stream으로 더욱 간결하게 나타낼 수는 있을 것 같습니다!

ASak1104 replied:
메서드 이름도 바꾸면 좋을 것 같아요
`get()` 도 좋을 것 같네요

shoeone96 replied:
모두 반영해서 수정했습니다!

IjjS commented on src/main/java/team/silvertown/masil/user/exception/OAuthValidator.java:
전부 null check인데 throwIf대신 notNull을 사용해도 될 것 같습니다

shoeone96 replied:
이게 notNull 할 때 BadRequestException 을 반환하는데 토큰관련된 Exception을 던져주는 게 더 적합할 것 같아서 throwIf로 처리했습니다!

IjjS commented on src/main/java/team/silvertown/masil/user/exception/InvalidAuthenticationException.java:
이 예외 위치는 common/exception에 들어가는게 좋을 것 같아요!

ASak1104 replied:
@Team-SilverTown/backend user랑 security 패키지로 분리해서 거기에 넣는 건 어때요?

IjjS commented on src/main/java/team/silvertown/masil/user/dto/LoginResponseDto.java:
```diff
@@ -0,0 +1,10 @@
+
+public record LoginResponseDto(
+    String token,
+    List<String> authority
```
로그인에 권한을 반환해서 응답에 내려줄 필요가 있다고 생각하신 이유가 있을까요?

ASak1104 replied:
중원형님 `authority` 토큰에 저장하는 방식은 어때요?

shoeone96 replied:
프론트에서도 처음 가입할 때 이걸 보고 추가 정보를 넣는 페이지로 갈지 아니면 바로 redirect 할지 결정할 수 있다고 생각했어요!

IjjS commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,67 @@
+        if (oauth2Authentication instanceof OAuth2AuthenticationToken oauth2Token) {
+            OAuth2User oAuth2User = oauth2Token.getPrincipal();
+            String provider = oauth2Token.getAuthorizedClientRegistrationId();
+            User user = processUserOAuth2UserJoin(oAuth2User, provider);
+            String jwtToken = tokenProvider.createToken(user.getId());
+
+            LoginResponseDto loginResponseDto = userService.login(jwtToken, user);
```
join -> login보다는 login -> 없으면 join이 낫지 않을까 싶습니다! 어떻게 생각하시는지요

shoeone96 replied:
로직을 좀 바뀔 것 같긴 한데 아래 내용 얘기하면서 같이 얘기해보면 좋을 것 같아요

IjjS commented on src/main/java/team/silvertown/masil/user/service/UserService.java:
저는 아무래도 카카오 로그인 시 로그인 시도 -> 없으면 회원가입 흐름이 더 낫다고 생각하는데 두 분은 어떻게 생각하시는지 궁금하네요. @ASak1104 @shoeone96 

현재는
회원가입 시 -> 유저 도메인 리턴
로그인 시 -> jwt, authorities 리턴
이렇게 되는데, 프론트에는 카카오로 시작하기 버튼 하나 밖에 없다고 생각합니다. 그럼 클라이언트에서 응답을 보고 회원가입인지 로그인인지 구분해야 하는데 유저 엔티티 전체를 내려주는 것과 authorities를 포함한 응답을 내려주는 것은 구분이 명확하지 않은 것 같고 과한 정보를 내려준다고 생각이 듭니다. 이 부분은 로그인 담당하시는 프론트와 이야기가 된 걸까요?

추가로 만약 지금 상태로 진행하게 된다면 메서드 구분도 되면 좋을 것 같습니다!!

생각하시느라 고민 많으셨을텐데 머리 아프게 해드리는 것 같아 죄송하네요..

ASak1104 replied:
> 저는 아무래도 카카오 로그인 시 로그인 시도 -> 없으면 회원가입 흐름이 더 낫다고 생각하는데 두 분은 어떻게 생각하시는지 궁금하네요.

이건 프론트랑 이야기해봐야할 것 같은데요?
로그인 시 마실가실 서버에 유저 정보가 없다면
바로 회원가입 페이지로 넘어가서 kakao api 호출 화면이 나오는 걸 생각하시는거죠?
저는 좋다고 생각하는데 프론트 분들에게 여쭤보고 시간을 좀 드려야겠네요

IjjS replied:
카카오로 로그인하는거면 로그인 때도 카카오 api 호출이 필요하지 않나요??

ASak1104 replied:
월요일에 한 번 부숴보시죠
팀 교류회 준비도 하고

shoeone96 replied:
음 아니면 기존 로직은 유지하되 회원가입이 되면 isSignup 같은 변수를 같이 내려줘서 이걸 true false로 구분해서 내려주는 것도 괜찮을 것 같아요

IjjS replied:
로그인보단 본인 정보 조회하는 api를 통해서 리다이렉트를 해도 될 것 같긴한데..담당하시는 프론트 분이랑 합의보시고 형님 하시고 싶은대로 하셔도 될 것 같아요!!

IjjS commented on src/main/java/team/silvertown/masil/user/util/CustomUserArgumentResolver.java:
```diff
@@ -0,0 +1,41 @@
+import org.springframework.web.method.support.HandlerMethodArgumentResolver;
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+public class CustomUserArgumentResolver implements HandlerMethodArgumentResolver {
```
지금은
1. JwtAuthenticationFilter에서 임시 UsernamePasswordAuthenticationToken 발급
2. requestMatcher를 이용해 AuthorizationFilter에서 인증인가 검증
3. 예외 발생 시 EntryPoint, AccessDeniedHandler

이런 프로세스인데 이러면 해당 resolver는 필요 없는 것 같습니다. UsernamePasswordAuthenticationToken을 발급하니 AuthenticationPrincipal로도 아이디를 뽑아낼 수 있으니까요. 만약 매처를 사용 안 하고 예외를 무조건 컨트롤러에서만 발생시켜 컨트롤러 어드바이스에서만 일괄적으로 처리하겠다면 resolver는 그런 의도를 달성시켜줄 수 있을 것 같습니다. 왜냐하면 Controller -> Resolver 에서 발생하는 예외는 어드바이스가 잡을 수 있을 것 같으니까요.

혹은 UsernamePasswordAuthenticationToken이 아니라 커스텀한 토큰이면 더더욱 쓸 필요가 있을 것 같습니다.

만약 UsernamePasswordAuthenticationToken을 사용하고 필터에서 인증인가를 거르고 싶다면 이 resolver는 필요없을 것 같고, UsernamePasswordAuthentication에서 했던 것처럼
1. JwtAuthenticationFilter에서 임시 UsernamePasswordAuthenticationToken 발급
2. 발급된 토큰으로 유저 서비스를 통해 principal이 유저인 UsernamePasswordAuthenticationToken 발급

하는 식으로도 할 수 있겠네요..!

IjjS commented on src/test/java/team/silvertown/masil/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,186 @@
+
+        //when
+        User joinedUser = userService.join(oAuth2User, VALID_PROVIDER);
+        userRepository.flush();
```
join이 transactional로 묶여있으니 flush를 하지 않아도 괜찮지 않나요??

IjjS commented on src/main/resources/application.yml:
👍 👍

IjjS reviewed (commented):
고생하셨습니다!! 고민 많이 되셨을건데 고민을 더 남겨드리는 기분이네요 ㅎㅎ..

IjjS commented on src/main/resources/application.yml:
```diff
@@ -13,5 +13,27 @@ spring:
+            client-id: ${CLIENT_ID}
+            client-secret: ${CLIENT_SECRET}
+            scope: profile_nickname
+            redirect-uri: "http://localhost:8080/login/oauth2/code/{registrationId}"
```
생각해보니까 이 부분도 키처리를 해야할 것 같은데요..?!

ASak1104 replied:
이것도 따로 분리를 해야겠네요

ASak1104 commented on .idea/codeStyles/codeStyleConfig.xml:
@Team-SilverTown/backend 형님들 이거 필요한가요?

IjjS replied:
.idea가 이그노어에 잘 안 들어갔나보다

ASak1104 commented on build.gradle:
```diff
@@ -42,8 +42,17 @@ dependencies {
     // faker
     implementation 'net.datafaker:datafaker:2.1.0'
+
 }
```
여기 개행 없어도 될 것 같은데요?

ASak1104 commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,90 @@
+@Component
+public class JwtTokenProvider {
+
+    public static final String USER_ID_CLAIM = "user_id";
+    public static final int MILLS = 1000;
+    private final long tokenValidityInMilliseconds;
```
형님 여기 개행이요~
그리고 이전 빽엔두리에서 저는 상수를 모두 퍼블릭으로 했었는데
개발할 때 공개 유무 방식 룰을 정해서 기본으로 가져갈 필요가 있을까 같은 것도 고민을 해봐야겠네요

shoeone96 replied:
전체적으로 돌아가는 게 문제가 없고 다른 곳에서 사용 안해서 private으로 사용할게요!

ASak1104 commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,90 @@
+        return true;
+    }
+
+    private void loggingException(RuntimeException e) {
+        if (e instanceof SecurityException || e instanceof MalformedJwtException) {
+            log.debug(TokenMessage.INVALID_SIGNATURE, e);
+        } else if (e instanceof ExpiredJwtException) {
+            log.debug(TokenMessage.EXPIRED_TOKEN, e);
+        } else if (e instanceof UnsupportedJwtException) {
+            log.debug(TokenMessage.UNSUPPORTED_TOKEN, e);
+        } else {
+            log.debug(TokenMessage.UNEXPECTED_TOKEN, e);
+        }
+    }
```
지금 생각해보면 `TokenMessage` 상수 클래스를 `TokenExceptionLogger` 로 바꿔서 해당 메서드를 로거로 옮기면 좋을 것 같아요!

shoeone96 replied:
음 내부에 static 상수로 선언만 하는 방식으로 바꿨는데 이건 이상할까요?

ASak1104 commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    private User processUserOAuth2UserJoin(OAuth2User oAuth2User, String provider) {
+        return userService.join(oAuth2User, provider);
+    }
```
형님 이거 따로 분리한 이유가 있나요??

shoeone96 replied:
분리 안하는 게 더 좋아보여서 그렇게 만들었어요!

ASak1104 commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+    @Bean
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+        http
+            .authorizeHttpRequests(auth -> {
+                auth.requestMatchers("/oauth2/**")
+                    .permitAll();
+                auth.anyRequest()
+                    .authenticated();
+            });
```
이 부분은 추후 코드 변경이 자주 일어날 것 같은데
따로 `Configuration`으로 빼는 건 어때용?

shoeone96 replied:
지난 번에 사용했던 클래스 가져와봤어요 👍 
확실히 이번엔 권한도 들어가는만큼 역할을 더 분리하는 게 맞는 것 같네요

ASak1104 commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+                handler -> handler.authenticationEntryPoint(customAuthenticationEntryPoint));
+
+        //기타 설정 +  OAuth successHandler 설정
+        http
```
이건 한 체인으로 묶는게 이뻐보이네요

ASak1104 commented on src/main/java/team/silvertown/masil/config/security/WebSecurityConfig.java:
```diff
@@ -0,0 +1,102 @@
+            Set<GrantedAuthority> mappedAuthorities = new HashSet<>();
+
+            authorities.forEach((authority) -> {
+                GrantedAuthority mappedAuthority;
+
+                if (authority instanceof OidcUserAuthority userAuthority) {
+                    mappedAuthority = new OidcUserAuthority(
+                        "OIDC_USER", userAuthority.getIdToken(), userAuthority.getUserInfo());
+                } else if (authority instanceof OAuth2UserAuthority userAuthority) {
```
`GrantedAuthority` 부분 메서드로 분리하면 너무 좋을 것 같아요

ASak1104 commented on src/main/java/team/silvertown/masil/user/exception/UserErrorCode.java:
```diff
@@ -0,0 +1,31 @@
+
+import team.silvertown.masil.common.exception.ErrorCode;
+
+public enum UserErrorCode implements ErrorCode {
+
+    INVALID_OAUTH2_TOKEN(10030000, "소셜로그인 과정에서 문제가 발생했습니다."),
```
enum에서 저희 안 띄우기로 하지 않았나요?

IjjS replied:
제 기억으론 entry point에서 썼던 것 같은데 enum error code로도 이렇게 관리할 수 있다면 나쁘진 않은 것 같습니다

ASak1104 replied:
<img width="299" alt="image" src="https://github.com/Team-SilverTown/Team-SilverTown-MasilGasil-BE/assets/31983961/affe1459-1daf-4a84-9b98-8e1b2654fa82">

행님 코드 6번 행에서의 개행 말씀드린겁니다!

ASak1104 commented on src/main/java/team/silvertown/masil/user/service/UserService.java:
```diff
@@ -0,0 +1,66 @@
+                userAuthorityRepository.save(newAuthority);
+                return savedUser;
+            });
+    }
```
여기는 early return 이랑 메서드 분리해주면 코드가 예쁘게 나올 것 같아요 형님

ASak1104 commented on src/main/java/team/silvertown/masil/user/service/UserService.java:
```diff
@@ -0,0 +1,66 @@
+        UserValidator.validateAuthority(userAuthorities);
+
+        List<String> authorities = userAuthorities.stream()
+            .map(authority -> authority.getAuthority()
+                .getAuthority())
```
`.getAuthority()`를 두 번 호출하니까 좀 그렇네용
어떻게 개선이 가능할까요?

shoeone96 replied:
SimpleGrantedAuthority(ROLE_PREFIX + authority.name());

이 부분이니까 getName()으로 만들었고 실제 사용하는 람다 메서드도 개선했습니당

ASak1104 commented on src/main/resources/application.yml:
```diff
@@ -13,5 +13,27 @@ spring:
+            client-authentication-method: client_secret_post
+        provider:
+          kakao:
+            authorization-uri: https://kauth.kakao.com/oauth/authorize
+            token-uri: https://kauth.kakao.com/oauth/token
+            user-info-uri: https://kapi.kakao.com/v2/user/me
+            user-name-attribute: id
```
이런 부분은 버저닝으로 인해 바뀔 수 있으니까 깃허브 variable로 관리하는 것도 괜찮겠네요

ASak1104 commented on src/main/resources/application.yml:
```diff
@@ -13,5 +13,27 @@ spring:
+          kakao:
+            client-name: kakao
+            client-id: ${CLIENT_ID}
+            client-secret: ${CLIENT_SECRET}
```
이런 이름의 키는 혼동을 줄 수 있으니까 나중을 생각해서 `KAKAO_` prefix를 주는건 어때요??

ASak1104 commented on src/main/java/team/silvertown/masil/user/util/CustomUserArgumentResolver.java:
@Team-SilverTown/backend 이 부분은 같이 이야기를 해봐야겠네요

ASak1104 commented on src/test/java/team/silvertown/masil/config/jwt/JwtTokenProviderTest.java:
```diff
@@ -0,0 +1,36 @@
+        //then
+        assertThat(jwtTokenProvider.validateToken(token)).isTrue();
+        assertThat(jwtTokenProvider.getAuthentication(token)
+            .getPrincipal() instanceof Long).isTrue();
```
이거 인스턴스 assert method가 있을거에요!

ASak1104 commented on src/test/java/team/silvertown/masil/config/jwt/JwtTokenProviderTest.java:
```diff
@@ -0,0 +1,36 @@
+        assertThat(jwtTokenProvider.validateToken(token)).isTrue();
+        assertThat(jwtTokenProvider.getAuthentication(token)
+            .getPrincipal() instanceof Long).isTrue();
+        Long decodedUserId = (Long) jwtTokenProvider.getAuthentication(token)
+            .getPrincipal();
+        assertThat(userId).isEqualTo(decodedUserId);
```
그냥 이 부분만 살려놓고 위 검증 메서드는 날려도 괜찮을 것 같은데요?

ASak1104 commented on src/test/java/team/silvertown/masil/user/domain/ProviderTest.java:
```diff
@@ -0,0 +1,38 @@
+        //given, when
+        String invalidProvider = "bug";
+
+        //then
+        Assertions.assertThatThrownBy(() -> Provider.returnProvider(invalidProvider))
+            .isInstanceOf(InvalidAuthenticationException.class)
```
인제 형님이 typeOf를 쓰셔서 그걸로 통일하는 것도 생각해봐야겠네요
저는 상관 없습니다~

ASak1104 commented on src/test/java/team/silvertown/masil/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,186 @@
+
+
+    @Test
+    public void 정상적으로_처음_회원가입하는_유저는_유저_정보를_저장한_후_유저객체를_return한다() throws Exception {
```
`유저객체를 반환한다`로 가시죠

ASak1104 commented on src/test/java/team/silvertown/masil/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,186 @@
+    @Test
+    public void 정상적으로_처음_회원가입하는_유저는_유저_정보를_저장한_후_유저객체를_return한다() throws Exception {
+        //given
+        authorityRepository.deleteAll();
+        userRepository.deleteAll();
```
@Team-SilverTown/backend 이거는 조금 생각을 해봐야겠는데요?
스크럼이나 주말에 잠깐 이야기 해보시죠!

ASak1104 reviewed (commented):
이야기할게 많은 것 같은데요?
형님들 시간 한 번 내시죠!

shoeone96 committed:
refactor: test-yaml 파일에서 secret 정보 삭제 후 config 파일로 설정

shoeone96 committed:
refactor: @DisplayNameGeneration 빠진 부분 수정

shoeone96 committed:
refactor: filter 중복 사용방지를 위한 OncePerRequestFilter 도입

기존 GenericFilterBean -> OncePerRequestFilter로 변경하여 redirect 등의 상황에서 중복으로 불필요한 필터 사용을 방지

shoeone96 committed:
refactor: springboot 버전에 맞게 OAuth2 버전 변경

shoeone96 committed:
refactor: bean에 등록된 ObjectMapper 사용으로 변경

shoeone96 committed:
style: 사용하지 않는 주석 제거

shoeone96 committed:
refactor: http 관련 설정 변경

.headers(AbstractHttpConfigurer::disable) 삭제
.requestCache(AbstractHttpConfigurer::disable) 추가

shoeone96 committed:
refactor: stream을 활용한 Provider 판별 메서드 리팩토링

사용하는 곳에서 메서드 명도 변경

shoeone96 committed:
refactor: security 폴더 분리

shoeone96 committed:
refactor: 필요없은 repository method 제거

shoeone96 committed:
chore: git ignore 추가

shoeone96 committed:
style: 개행 삭제

shoeone96 committed:
refactor: 클래스 내 사용하는 상수 private 접근 제어자로 변경

shoeone96 committed:
refactor: token message를 token provider 내부의 상수로 선언

shoeone96 committed:
refactor: 불필요한 메서드 분리 삭제

shoeone96 committed:
refactor: http method 분리

shoeone96 committed:
style: 하나의 메서드 체이닝에 통합

shoeone96 committed:
refactor: 메서드 분리

shoeone96 committed:
refactor: method 명 수정 및 람다 식 간결하게 변경

shoeone96 committed:
style: clientId 구분자로 prefix 추가

shoeone96 committed:
refactor: type 검증 assert문으로 변경

shoeone96 committed:
refactor: 동일 내용 검증 메서드 삭제

shoeone96 committed:
style: 테스트 명 변경

shoeone96 committed:
fix: conflict fixed

IjjS reviewed (approved):
프론트 결정되면 반영주세요!

shoeone96 committed:
refactor: login response에 토큰만 주는 방식으로 변경

shoeone96 committed:
refactor: 사용하지 않는 로직 삭제 및 private method 위치 변경

shoeone96 committed:
feat: 토큰 생성 시 유저 권한 같이 넣어주는 로직 추가

shoeone96 committed:
feat: user response 내려줄 때 권한 넣는 로직 삭제

shoeone96 committed:
test: User 생성 시 권한도 명확히 들어가 있는지 확인

shoeone96 committed:
새로 생성된 유저에 대해 토큰에 권한도 같이 들어가는지 확인하는 로직 추가

shoeone96 commented:
@Team-SilverTown/backend 

원주님과 얘기해본 결과 /me 호출 시 권한을 확인하는 과정에서 반려당하면 권환 확인이 안되면 돌아가는 로직으로 결정했습니다.
1. LoginResponse에서 토큰을 확인하는 로직을 제거했습니다.
2. 토큰에 authority를 부여하는 로직을 추가하여 가지고 있는 권한을 모두 추가하도록 하였습니다.

ASak1104 commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
```diff
@@ -33,6 +41,9 @@ public class JwtTokenProvider {
     private final JwtParser jwtParser;
 
+    @Autowired
+    private UserAuthorityRepository userAuthorityRepository;
```
이 부분은 `final` 키워드가 들어가면 좋겠네요
필드 injection 보다는 상위 클래스에서 해당 레포를 주입해주고
생성자에 repo를 넣어주는 방식으로 하는게 더 좋아보여요

ASak1104 commented on src/main/java/team/silvertown/masil/user/repository/UserAuthorityRepository.java:
```diff
@@ -2,11 +2,14 @@
 public interface UserAuthorityRepository extends JpaRepository<UserAuthority, Long> {
 
     List<UserAuthority> findByUser(User user);
+    @Query(value = "select authority from UserAuthority authority where authority.user.id = :userId")
```
SQL 쿼리 대문자로 통일하고
테이블명 snake case로 쓰는건 어때요?

IjjS replied:
jpa 쿼리라서 테이블명 대신 클래스명이 들어가는게 아닌가요??

ASak1104 replied:
아 그렇네 JPQL로 작성하니까 테이블 명으로 하면 안되겠네요!!
그런 이유라면 그냥 소문자로 쿼리 쓰고 사용하는게 확실히 낫겠어요

ASak1104 commented on .idea/codeStyles/codeStyleConfig.xml:
`.gitignore`에 추가했으니
해당 파일 내려서 커밋해주시면 될 것 같아요!

ASak1104 commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,112 @@
+    private static final String UNEXPECTED_TOKEN = "JWT token compact of handler are invalid.";
+    private static final String USER_ID_CLAIM = "user_id";
+    private static final int MILLS = 1000;
+    private final long tokenValidityInMilliseconds;
```
여기 개행이 빠졌네요!

ASak1104 commented on src/main/java/team/silvertown/masil/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,112 @@
+        return true;
+    }
+
+    private void loggingException(RuntimeException e) {
+        if (e instanceof SecurityException || e instanceof MalformedJwtException) {
+            log.debug(INVALID_SIGNATURE, e);
+        } else if (e instanceof ExpiredJwtException) {
+            log.debug(EXPIRED_TOKEN, e);
+        } else if (e instanceof UnsupportedJwtException) {
+            log.debug(UNSUPPORTED_TOKEN, e);
+        } else {
+            log.debug(UNEXPECTED_TOKEN, e);
+        }
+    }
```
이 부분 분리를 해야하나....
아니면 지워야하나...
저번에 언급을 했던 것 같은데 중원 형님께 맡기겠습니다

shoeone96 replied:
음 분리가 많이 필요한 느낌인가요? 그대로도 괜찮을 것 같긴 해서 고민이 되네요

ASak1104 commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,58 @@
+
+@Slf4j
+@RequiredArgsConstructor
+public class OAuth2AuthenticationSuccessHandler extends
+    SavedRequestAwareAuthenticationSuccessHandler {
```
`extends` 키워드 내리는게 좋을 것 같은데요?

ASak1104 commented on src/main/java/team/silvertown/masil/config/oauth2/OAuth2AuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,58 @@
+    private final JwtTokenProvider tokenProvider;
+    private final ObjectMapper objectMapper;
+
+    @Override
+    public void onAuthenticationSuccess(
+        HttpServletRequest request, HttpServletResponse response,
+        Authentication oauth2Authentication
```
```suggestion
    @Override
    public void onAuthenticationSuccess(
        HttpServletRequest request,
        HttpServletResponse response,
        Authentication oauth2Authentication
```

개행하는 김에 다 하는건 어때요?

ASak1104 commented on src/test/java/team/silvertown/masil/JwtTestConfig.java:
```diff
@@ -0,0 +1,21 @@
+public class JwtTestConfig {
+
+    private static final String JWT_ISSUER = "test issuer";
+    private static final String BASE64_SECRET = "testSecrettestSecrettestSecrettestSecrettestSecrettestSecrettestSecrettestSecrettestSecrettestSecret";
```
제 기억 상 jjwt에서 secret 키를 generate 할 수 있었던 것 같은데
테스트 코드여도 개선하면 좋을 것 같습니다.
임시 방편으로 String `repeat`으로 코드를 보기 좋게하는것도 좋을 것 같고요

그냥 test yaml 파일에 값을 넣어놔도 저는 큰 상관 없다고 생각하긴 합니다.

ASak1104 commented on src/test/java/team/silvertown/masil/JwtTestConfig.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.context.annotation.Configuration;
+import team.silvertown.masil.config.jwt.JwtProperties;
+
+@Configuration
+public class JwtTestConfig {
```
`@TestConfiguration`으로 바꿔야할 것 같아요!

shoeone96 replied:
이거 @TestConfiguration으로 하면 못읽었어가지고 좀 더 레퍼런스 보고 수정해볼게요

ASak1104 commented on src/test/java/team/silvertown/masil/OAuth2TestConfig.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+
+@Configuration
+public class OAuth2TestConfig {
```
```suggestion
@TestConfiguration
public class OAuth2TestConfig {
```

ASak1104 commented on src/test/java/team/silvertown/masil/OAuth2TestConfig.java:
```diff
@@ -0,0 +1,37 @@
+        return new InMemoryClientRegistrationRepository(clientRegistration());
+    }
+
+    @Bean
+    public ClientRegistration clientRegistration() {
+        return ClientRegistration.withRegistrationId("kakao")
+            .clientId("test-client-id")
+            .clientSecret("test-client-secret")
+            .scope("profile_nickname")
+            .authorizationUri("https://kauth.kakao.com/oauth/authorize")
+            .tokenUri("https://kauth.kakao.com/oauth/token")
+            .userInfoUri("https://kapi.kakao.com/v2/user/me")
+            .userNameAttributeName("id")
+            .clientName("kakao")
+            .redirectUri("http://localhost:8080/login/oauth2/code/kakao")
+            .clientAuthenticationMethod(ClientAuthenticationMethod.CLIENT_SECRET_POST)
+            .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
+            .build();
+    }
```
지금 방식에서는 test yaml 파일에 넣어서 자동으로 빈을 만드는게 나아보여요!
특히 redirectUri 중 포트 같은 건 테스트 환경에서 달라질 수 있으니까 신경을 써야겠네요

그리고 프로덕션 외부 api를 테스트에서 사용하는 건 문제가 있을 수 있어서
test api를 찾거나 mocking을 기가 막히게 해야겠네요

IjjS replied:
사실 이걸 테스트 때 사용하는게 아니라 빈 등록만이 문제라서 아예 똥값만 줘서 사용 안 한다는 걸 명시해도 괜찮을 것 같습니다

ASak1104 replied:
음~ 좋네요!
그냥 더미 데이터를 넣는 식으로 하거나 주석 같은 걸로 명시하는 걸 말씀하시는거죠!
test api 적용은 차차 같이 알아봐요!

shoeone96 replied:
맞아요 springboot 시작 시 bean에서 안터지게 만들려고 한 거라 나중에 redirecturi 가 사용되는 경우가 생기면 모킹 해볼게요!

ASak1104 commented on src/test/java/team/silvertown/masil/user/domain/ProviderTest.java:
```diff
@@ -0,0 +1,41 @@
+        String invalidProvider = "bug";
+
+        //then
+        Assertions.assertThatThrownBy(() -> Provider.get(invalidProvider))
+            .isInstanceOf(InvalidAuthenticationException.class)
+            .hasMessage(UserErrorCode.INVALID_PROVIDER.getMessage());
```
이왕 하는거 이 부분도 static import 문으로 바꾸죠!
```suggestion
        assertThatThrownBy(() -> Provider.get(invalidProvider))
            .isInstanceOf(InvalidAuthenticationException.class)
            .hasMessage(UserErrorCode.INVALID_PROVIDER.getMessage());
```

ASak1104 reviewed (commented):
흠 항상 고민할 게 많네요
의논하다가 머지가 안되겠어요 ㅎㅎ

shoeone96 committed:
refactor: 생성자 주입 방식으로 UserAuthorityRepository 멤버로 넣기

shoeone96 committed:
style: static method import

shoeone96 committed:
style: 개행 수정

shoeone96 committed:
refactor: @TestConfiguration 적용

@Configuration -> @TestConfiguration

shoeone96 committed:
refactor: @TestConfiguration @Import로 적용

shoeone96 committed:
refactor: @TestConfiguration -> @Configuration

main 실행 함수에서 @TestConfiguration을 읽지 못하는 문제로 rollback

shoeone96 committed:
refactor: 사용하지 않는 상수 삭제

shoeone96 committed:
style: ignore 추가한 파일 삭제

shoeone96 committed:
Merge branch 'feature' into feature-login

ASak1104 reviewed (approved):
이렇게 길어지면 안되겠죠 얼른 머지하고 리팩토링하는 걸로 합의!
